File: idle-kluch-client/src/ecs/systems/ResourceTransportSystem.js
import { STORAGE_NODE } from "../nodes";
import { ResourceStorageComponent } from "../components/ResourceStorageComponent";
import { distance } from "../utils";

export class ResourceTransportSystem {

	constructor(engine) {
		this.engine = engine;
		this.node = STORAGE_NODE;
	}

	getEngine() {
		return this.engine;
	}

	shouldUpdate = () => true;

	update = (delta) => {
		this.getEngine()
			.getEntitiesByNode(this.node)
			.forEach(entity => this.updateEntity(entity, delta));
	}

	updateEntity = (entity, delta) => {
		const storage = entity.getComponent(ResourceStorageComponent);
		if (storage.routes.length < storage.maxRoutes) {
			storage.nextRouteProgress = Math.min(storage.secondsPerRoute, storage.nextRouteProgress + delta);
		}

		storage.getRoutes().forEach(route => {
			if (route.finished) {
				return storage.removeRoute(route);
			}

			const target = this.getEngine().getEntityById(route.to);
			if (!target) {
				return;
			}
			const distanceBetweenEntities = distance(entity, target);
			const distanceChange = storage.transportSpeed * delta;
			const progressChange = distanceChange / distanceBetweenEntities;
			route.progress = Math.min(1, route.progress + progressChange);

			const targetStorage = target.getComponent(ResourceStorageComponent);
			if (route.progress >= 0.99 && targetStorage.hasRemainingCapacity()) {
				targetStorage.addResource(route.resource);
				route.finished = true;
			}

		});
	}

}

File: idle-kluch-server/src/main/java/com/soze/idlekluch/game/engine/systems/ResourceTransportSystem.java
package com.soze.idlekluch.game.engine.systems;

import com.soze.idlekluch.core.utils.jpa.EntityUUID;
import com.soze.idlekluch.game.engine.EntityUtils;
import com.soze.idlekluch.game.engine.components.ResourceStorageComponent;
import com.soze.idlekluch.game.engine.components.resourcetransport.ResourceRoute;
import com.soze.idlekluch.game.engine.nodes.Nodes;
import com.soze.klecs.engine.Engine;
import com.soze.klecs.entity.Entity;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Iterator;
import java.util.List;
import java.util.Map;


public class ResourceTransportSystem extends BaseEntitySystem {

  private static final Logger LOG = LoggerFactory.getLogger(ResourceTransportSystem.class);

  private static final float METERS_PER_SECOND = 30f;

  public ResourceTransportSystem(final Engine engine,
                                 final Map<EntityUUID, Entity> changedEntities) {
    super(engine, changedEntities);
  }

  @Override
  public void update(final float delta) {
    getStorages().forEach(entity -> update(entity, delta));
  }

  private void update(final Entity entity, final float delta) {
    final ResourceStorageComponent storage = entity.getComponent(ResourceStorageComponent.class);
    final Iterator<ResourceRoute> it = storage.getRoutes().iterator();
    while (it.hasNext()) {
      final ResourceRoute route = it.next();
      final Entity target = getEngine().getEntityById(route.getTo()).get();
      final float distance = EntityUtils.distance(entity, target);
      final float progress = route.getProgress();
      final float distanceChange = METERS_PER_SECOND * delta;
      final float progressChange = distanceChange / distance;
      final float nextProgress = progress + progressChange;
      route.setProgress(Math.min(1f, nextProgress));

      final ResourceStorageComponent targetStorage = target.getComponent(ResourceStorageComponent.class);
      if(nextProgress >= 0.99f && targetStorage.hasRemainingCapacity()) {
        LOG.trace("Finished route [{}]", route);
        targetStorage.addResource(route.getResource());
        it.remove();
      }
    }
  }

  private List<Entity> getStorages() {
    return getEngine().getEntitiesByNode(Nodes.STORAGE);
  }

}
